### PR TITLE
Adds support for mj-font, mj-preview and mj-raw 

### PR DESCRIPTION
**Pull request by barsch**
tackles issue #7 - all tags have at least one test case - ```mj-raw``` is a bit ugly but couldn't find a better solution

May I suggest to use [black](https://github.com/psf/black) as automatic code formatter - it was hard for me to emulate your code style ;)

### CODE REVIEW

**Review by FelixSchwarz — CHANGES_REQUESTED** [review]
This commit could use a better commit message. Basically your issue text from https://github.com/FelixSchwarz/mjml-stub/issues/7#issue-789196342 with slight adaptation would be fine.

**Review by FelixSchwarz — CHANGES_REQUESTED** [review]
This commit could use a better commit message. Basically your issue text from https://github.com/FelixSchwarz/mjml-stub/issues/7#issue-789196342 with slight adaptation would be fine.

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,5 @@
 /dist
 /*.sh
 /venv*
-
```
Comment (FelixSchwarz) [inline .gitignore:5]: I often leave a single trailing newline because GTK-based editors on Fedora 33+ sometimes do not show the last line - but no hard feelings about this. So far this was just me fixing a client's problem. I can adapt when others start contributing.

File: mjml/mjml2html.py [hunk 2]
```diff
@@ -113,6 +114,11 @@ def default_attr_classes(value):
             )
 
             _parse_mjml = lambda mjml: parse(mjml, nextParentMjClass)
+            # XXX: ugly - but need raw XML content for mj-raw tag
+            if tagName == 'mj-raw':
+                from lxml.etree import tostring
+                content = tostring(_mjml).strip().decode('utf-8')
+                content = content.replace('<mj-raw>', '').replace('</mj-raw>', '')
```
Comment (FelixSchwarz) [inline mjml/mjml2html.py:121]: Not sure what `lxml.etree.tostring()` will return but this seems very specific. Does it handle something like `<mj-raw >`? (extra space)

Also this is basically about stripping the `<mj-raw>...</mj-raw>` container, right? Is there no functionality directly in lxml which we could use here?

**Review by FelixSchwarz — COMMENTED** [review]
Thank you very much for your contribution. Pretty neat, just a few stylistic notes which I added via github's interface (which I still find highly confusing as I usually review code changes for each commit separately).

Two more issues:

- Please squash the changes to `.gitignore`. These are really trivial. Often I just write something like ".gitignore: ignore eclipse project and .pyc files".
- "fix not working error message" needs a better commit message. Basically your issue text from #7 (comment) with slight adaptation would be fine.

If you start using the Python port in production I'd be very interested if you see any issues/what should be improved (and maybe what works great). Also how did you came across this project?

> May I suggest to use black as automatic code formatter - it was hard for me to emulate your code style ;)

Well, so far this was just a quick thing for one person to solve a customer's problem so I did not care much about a fixed coding style as it might be shared in the Python community. However if this becomes an actual open source project with multiple contributors I guess I'll have to adapt a bit :-)

File: mjml/mjml2html.py [hunk 3]
```diff
@@ -113,6 +114,11 @@ def default_attr_classes(value):
             )
 
             _parse_mjml = lambda mjml: parse(mjml, nextParentMjClass)
+            # XXX: ugly - but need raw XML content for mj-raw tag
+            if tagName == 'mj-raw':
+                from lxml.etree import tostring
```
[the remark this answers is not in the document]
Comment (barsch) [reply mjml/mjml2html.py:119]: this import is only needed for this mess of code - so if we find a better way to extract the unmodified content between the starting and closing ```<mj-raw>``` tags those few lines are not needed anymore

File: mjml/mjml2html.py (hunk 2 above)
Comment (barsch) [reply mjml/mjml2html.py:121]: > Not sure what `lxml.etree.tostring()` will return but this seems very specific. Does it handle something like `<mj-raw >`? (extra space)

the ```.strip()```after could be removed - on the other side extra space shouldn't matter in the final HTML document

> Also this is basically about stripping the `<mj-raw>...</mj-raw>` container, right? Is there no functionality directly in lxml which we could use here?

correct, afaik there is no such function within lxml  - but there are multiple approaches to solve it:

https://stackoverflow.com/questions/4624062/get-all-text-inside-a-tag-in-lxml
https://stackoverflow.com/questions/11122397/how-do-i-get-the-whole-content-between-two-xml-tags-in-python

I don't mind which one to use if we have to - up to you to decide - and thats the reason the whole paragraph has a ```XXX: ugly``` above ;)

File: .gitignore (hunk 1 above)
Comment (barsch) [reply .gitignore:5]: that is actually strange - locally my .gitignore file has a trailing newline - I'm confused ...

```
C:\Users\barsch\git\mjml-stub>git status
On branch more_tags
Your branch is up to date with 'origin/more_tags'.

nothing to commit, working tree clean

C:\Users\barsch\git\mjml-stub>more .gitignore
/build
/dist
/*.sh
/venv*
/.project
*.pyc

C:\Users\barsch\git\mjml-stub>
```

**Comment by barsch** [comment]
> Two more issues:
> 
>     * Please squash the changes to `.gitignore`. These are really trivial. Often I just write something like ".gitignore: ignore eclipse project and .pyc files".
> 
>     * "fix not working error message" needs a better commit message. Basically your issue text from #7 (comment) with slight adaptation would be fine.

done


> If you start using the Python port in production I'd be very interested if you see any issues/what should be improved (and maybe what works great). Also how did you came across this project?

I hate to have a node based server running just for generating my newsletters - I tested it so far with our newsletters - not looking very good atm because of missing inline styles and the small issues I reported


> May I suggest to use black as automatic code formatter - it was hard for me to emulate your code style ;)
> Well, so far this was just a quick thing for one person to solve a customer's problem so I did not care much about a fixed coding style as it might be shared in the Python community. However if this becomes an actual open source project with multiple contributors I guess I'll have to adapt a bit :-)

the nice thing about automated code formatters is you don't have to adapt - you save your python file and its just does it magic - sure one needs to live with the style its generates but at least it completely eliminates any style discussions because its the same for everyone - I basically don't care about formatting at all anymore which took so much time away from actual programming especially if it is a multi user project ...

File: mjml/mjml2html.py (hunk 2 above)
Comment (FelixSchwarz) [reply mjml/mjml2html.py:121]: Thank you for the pointers. Not sure which version is easier to grasp. Also the comments mention a few edge-case bugs so we should definitively create a small utility function which we can test independently of the full mjml machinery.

I'll merge this PR as-is (it enhances the status quote and the part above is not too bad imho) and create ticket for this specific issue.

File: mjml/mjml2html.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (FelixSchwarz) [reply mjml/mjml2html.py:119]: fair enough - though I think in general we should place imports at the start of the file.

File: .gitignore (hunk 1 above)
Comment (FelixSchwarz) [reply .gitignore:5]: well, not that important I guess.

**Comment by FelixSchwarz** [comment]
As [mentioned in the review comments](https://github.com/FelixSchwarz/mjml-stub/pull/8#discussion_r561268014) I'll merge this PR as-is. The PR enhances the status quo and the "innerXML part" should not prevent merging this.

Again: Thank you very much for your contributions.

> I tested it so far with our newsletters - not looking very good atm because of missing inline styles and the small issues I reported

Do you plan to implement inline styles next? Do you need `<mj-style inline="inline">` or generic inlining? (or both)

**Review by FelixSchwarz — APPROVED** [review]
